### PR TITLE
[TN-4] [TN-3] Auth for Kafka and ES

### PR DESCRIPTION
**Pull request by tbuchaillot**
## Description
Adding configuration for auth mechanism in Kafka and ElasticSearch:

-  basic auth (plain and scram) for Kafka
-  basic auth (username and password) and API Key auth (api_key and api_key_id) for ElasticSearch.

**ElasticSearch:**
Basic auth configuration:
The fields **auth_basic_username** and **auth_basic_password** are added into ES configuration. Those fields are going to be send in the Authorization header as `username:password` encoded in base64. 
Example config:
```
"elasticsearch": {
      "type": "elasticsearch",
      "meta": {
        "index_name": "tyk_analytics",
        "elasticsearch_url": "https://elastichost:9243",
        "enable_sniffing": false,
        "document_type": "tyk_analytics",
        "rolling_index": false,
        "extended_stats": false,
        "version": "6",
        "auth_basic_username":"elastic",
        "auth_basic_password":"password"
      }
    }
```
API Key auth configuration:
The fields **auth_api_key_id** and **auth_api_key** are added into ES configuration. Those fields are going to be send in the Authorization header as  `ApiKey  base64(auth_api_key_id:auth_api_key)`.
Example config:
```
"elasticsearch": {
      "type": "elasticsearch",
      "meta": {
        "index_name": "tyk_analytics",
        "elasticsearch_url": "https://elastichost:9243",
        "enable_sniffing": false,
        "document_type": "tyk_analytics",
        "rolling_index": false,
        "extended_stats": false,
        "version": "6",
        "auth_api_key_id":"8bXvD3IB-IBDfS7oZEWV",
        "auth_api_key":"LRmuicydSW-zeb0Zd_3DGw"      
      }
    }
```
**Kafka:**
The fields **sasl_mechanism**, **sasl_algorithm**, **sasl_username** and **sasl_password** are added into Kafka configuration. 

- **sasl_mechanism**: It's the authentication mechanism. It could be `plain` or `scram`.
- **sasl_algorithm**: It's the algorithm specified for `scram` mechanism. It could be `sha-512` or `sha-256`, `sha-256` it's the default value if the algorithm is not configured.
- **sasl_username**: It's the username or token / key.
- **sasl_password**: It's the password or secret.

Example config for `plain` mechanism:
```
    "kafka": {
      "type": "kafka",
      "meta": {
        "broker": [
            "pkc-4yyd6.us-east1.gcp.confluent.cloud:9092"
        ],
        "client_id": "lkc-9wmd7",
        "topic": "test-topic",
        "timeout": 60,
        "compressed": true,
        "meta_data": {
            "key": "value"
        },
        "use_ssl":true,
        "sasl_mechanism":"plain",
        "sasl_username":"usernameD7SWRYRIQEFEXHNB",
        "sasl_password":"passwordD7SWRYRIQEFEXHNB"
      }
    }
```
Example config for `scram` mechanism:
```
    "kafka": {
      "type": "kafka",
      "meta": {
        "broker": [
            "testtest.confluent.cloud:9092"
        ],
        "client_id": "test-9wmd7",
        "topic": "test-topic",
        "timeout": 60,
        "compressed": true,
        "meta_data": {
            "key": "value"
        },
        "use_ssl":true,
        "sasl_mechanism":"scram",
        "sasl_algorithm":"sha-512",
        "sasl_username":"usernameD7SWRYRIQEFEXHNB",
        "sasl_password":"passwordD7SWRYRIQEFEXHNB"
      }
    }
```

## Related Issue
Issue #246

## Motivation and Context
Issue #246. Enterprise clients / SaaS.

## How This Has Been Tested
ElasticSearch:
  - basic auth: Tested with https://cloud.elastic.co/home and ES 5 / 6.
  - API Key auth: Test with https://cloud.elastic.co/home and ES 5 / 6 generating the API Key following https://www.elastic.co/guide/en/elasticsearch/reference/6.8/security-api-create-api-key.html.

Kafka:
Tested with Kafka cluster in https://confluent.cloud/.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [X] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [x] My change requires a change to the documentation.
- [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [X] Check your code additions will not fail linting checks:
  - [X] `go fmt -s`
  - [X] `go vet`

### CODE REVIEW

File: pumps/kafka.go [hunk 1]
```diff
@@ -61,11 +70,33 @@ func (k *KafkaPump) Init(config interface{}) error {
...
+		if k.kafkaConf.Algorithm == "sha-512" || k.kafkaConf.Algorithm == "SHA-512" {
+			algorithm = scram.SHA512
+		}
+		var mechErr error
+		mechanism, mechErr = scram.Mechanism(algorithm, k.kafkaConf.Username, k.kafkaConf.Password)
+		if mechErr != nil {
+			k.log.Fatal("Failed initialize kafka mechanism  : ", mechErr)
+		}
+	default:
+		k.log.Warn("Tyk pump doesn't support " + k.kafkaConf.SASLMechanism + " mechanism.")
```
Comment (letzya) [inline pumps/kafka.go:91]: Can we please add the word "sasl" to the message? for instance:
`k.log.WithField("SASL-Mechanism", k.kafkaConf.SASLMechanism).Warn("Tyk pump doesn't support this SASL mechanism"")`

File: pumps/kafka.go [hunk 2]
```diff
@@ -61,11 +70,33 @@ func (k *KafkaPump) Init(config interface{}) error {
...
+	}
+
```
Comment (letzya) [inline pumps/kafka.go:93]: Can we add a warning message if we have mechanism set AND tlsConfig not set?

File: pumps/kafka.go (hunk 1 above)
Comment (bmonteiro) [reply pumps/kafka.go:91]: Why not supporting sasl? That is the main auth type used in enterprise clients.

File: pumps/kafka.go (hunk 2 above)
Comment (tbuchaillot) [reply pumps/kafka.go:93]: sure!

File: pumps/kafka.go (hunk 1 above)
Comment (tbuchaillot) [reply pumps/kafka.go:91]: It's going to default only when the SASL Mechanism isn't **plain** or **scram**.

Comment (bmonteiro) [reply pumps/kafka.go:91]: But default is actually the main need for enterprise clients and it looks like it is not supported. Maybe something interesting here https://github.com/confluentinc/confluent-kafka-go ?

Comment (buger) [reply pumps/kafka.go:91]: @bmonteiro based on your example current PR covers your scenario. Just set:
```
"sasl_mechanism": "plain",
"sasl_username": "token",
"sasl_password": "api-key-secret"
```

File: pumps/elasticsearch.go [hunk 3]
```diff
@@ -38,6 +38,8 @@ type ElasticsearchConf struct {
 	Version            string                  `mapstructure:"version"`
 	DisableBulk        bool                    `mapstructure:"disable_bulk"`
 	BulkConfig         ElasticsearchBulkConfig `mapstructure:"bulk_config"`
+	Username           string                  `mapstructure:"basic_auth_username"`
+	Password           string                  `mapstructure:"basic_auth_password"`
```
[the remark this answers is not in the document]
Comment (tbuchaillot) [reply pumps/elasticsearch.go]: Added `auth_api_key_id` and `auth_api_key` to the config file for ApiKey auth.

**Comment by tbuchaillot** [comment]
Updated PR description

**Pull request merged**